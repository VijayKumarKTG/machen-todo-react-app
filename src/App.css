@import url('https://fonts.googleapis.com/css2?family=Josefin+Sans:wght@400;500;600;700&display=swap');
@import url('https://fonts.googleapis.com/css2?family=Pacifico&display=swap');

* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  font-family: 'Josefin Sans', sans-serif;
  text-align: center;
}

ul {
  list-style-type: none;
}

a {
  color: gold;
  text-decoration: none;
}

header {
  padding: 20px 30px;
  display: flex;
  justify-content: space-between;
  position: fixed;
  top: 0;
  left: 0;
  background-color: #ffffff;
  width: 100%;
  z-index: 2;
}

.brand {
  font-family: 'Pacifico', cursive;
  font-size: 20px;
  padding: 10px;
  border: 2px solid gold;
  cursor: pointer;
}

.username {
  justify-self: center;
  align-self: center;
}

.filters {
  display: flex;
  justify-content: space-evenly;
  margin-top: 20px;
}

.filters button {
  color: #000;
  border: 2px solid gold;
  padding: 5px 20px;
  outline: none;
  background: #ffffff;
  cursor: pointer;
  font-family: 'Josefin Sans', sans-serif;
  font-weight: bold;
}

.list {
  padding: 30px 50px;
  margin-top: 50px;
}

.list__item {
  text-align: left;
  margin: 20px 0px;
}

.list__text {
  width: 70%;
}

.checkbox {
  margin: auto 30px auto 0px;
}

.checkbox input {
  width: 15px;
  height: 15px;
}

.border__bottom {
  width: 0;
  border-bottom: 2px solid gold;
}

.list__item:hover .border__bottom {
  animation: borderwidth 0.7s ease-in;
  width: 100%;
}

@keyframes borderwidth {
  0% {
    width: 30px;
  }
  50% {
    width: 50%;
  }
  100% {
    width: 100%;
  }
}

.list__layout {
  display: flex;
}

.list__icons {
  justify-self: center;
  align-self: center;
}

.list__item__desc {
  margin: 5px 0px;
  font-size: 15px;
}

.icon {
  color: gold;
  cursor: pointer;
}

.icon:nth-of-type(odd):hover {
  font-weight: bold;
  color: rgba(0, 255, 0, 0.8);
}

.icon:nth-of-type(even):hover {
  font-weight: bold;
  color: rgba(255, 0, 0, 0.8);
}

.complete {
  text-decoration: line-through;
  color: gray;
  opacity: 0.5;
}

.plus,
.top {
  position: fixed;
  bottom: 10px;
  padding: 5px;
  border: 2px gold solid;
  border-radius: 50%;
  color: gold;
  cursor: pointer;
}

.plus {
  right: 10px;
}

.top {
  left: 10px;
}

#top {
  margin-top: 50px;
  margin-bottom: 70px;
}

.alert {
  position: absolute;
  z-index: 4;
  width: 80%;
  left: 10%;
  top: 30%;
  background: #fff;
  padding: 30px;
  border: 2px solid gold;
  animation: alert 1.5s ease-in-out;
}

@keyframes alert {
  0% {
    opacity: 0;
  }
  30% {
    opacity: 0.3;
  }
  50% {
    opacity: 0.5;
  }
  70% {
    opacity: 0.7;
  }
  100% {
    opacity: 1;
  }
}

.alert__title {
  font-weight: bold;
  font-size: 17px;
}

.alert__input {
  outline: none;
  border: 2px solid gold;
  width: 80%;
  margin: 10px auto;
  padding: 5px;
  text-align: center;
  font-family: 'Josefin Sans', sans-serif;
  font-weight: bold;
  display: block;
}

.alert__button {
  padding: 5px 10px;
  border: 2px solid gold;
  outline: none;
  background: #fff;
  cursor: pointer;
}

.alertActive {
  z-index: 3;
  background-color: rgba(128, 128, 128, 0.5);
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.toggleActive {
  display: none;
}
