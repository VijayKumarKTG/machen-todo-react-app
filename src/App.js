import React from 'react';
import './App.css';
import {
  faPlus,
  faChevronUp,
  faStickyNote,
} from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Todo } from './Todo';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      tempName: '',
      title: '',
      desc: '',
      todos: [],
      addTodoActive: false,
      updateTodoActive: false,
      filter: 'all',
    };
    this.onNameChange = this.onNameChange.bind(this);
    this.onNameSubmit = this.onNameSubmit.bind(this);
    this.onTodoChange = this.onTodoChange.bind(this);
    this.onTodoSubmit = this.onTodoSubmit.bind(this);
    this.activateAddTodo = this.activateAddTodo.bind(this);
    this.activateUpdateTodo = this.activateUpdateTodo.bind(this);
    this.onSubmitUpdateTodo = this.onSubmitUpdateTodo.bind(this);
    this.deleteTodo = this.deleteTodo.bind(this);
    this.onFilterAll = this.onFilterAll.bind(this);
    this.onFilterActive = this.onFilterActive.bind(this);
    this.onFilterCompleted = this.onFilterCompleted.bind(this);
    this.onComplete = this.onComplete.bind(this);
  }

  onNameChange(e) {
    this.setState({ tempName: e.target.value });
  }

  onNameSubmit() {
    this.setState({ name: this.state.tempName });
  }

  onTodoChange(e) {
    let value = e.target.value;
    let name = e.target.name;
    this.setState({ [name]: value });
  }

  onTodoSubmit(e) {
    e.preventDefault();
    this.setState((prevState) => {
      if (prevState.title && prevState.desc) {
        let list = prevState.todos;
        let id = list.length > 0 ? list[list.length - 1].id + 1 : 1;
        let todos = [
          ...list,
          {
            id,
            title: prevState.title,
            desc: prevState.desc,
            completed: false,
          },
        ];
        return {
          todos,
          title: '',
          desc: '',
          addTodoActive: false,
        };
      } else {
        return prevState;
      }
    });
  }

  activateAddTodo() {
    this.setState({ addTodoActive: true });
  }

  activateUpdateTodo(todo) {
    this.setState({
      title: todo.title,
      desc: todo.desc,
      updateTodoActive: true,
      i: todo.i,
    });
  }

  onSubmitUpdateTodo() {
    this.setState((prevState) => {
      let todos = [
        ...prevState.todos.slice(0, prevState.i),
        {
          id: prevState.todos[prevState.i].id,
          title: prevState.title,
          desc: prevState.desc,
          completed: prevState.todos[prevState.i].completed,
        },
        ...prevState.todos.slice(prevState.i + 1),
      ];
      return {
        todos,
        updateTodoActive: false,
        i: null,
        title: '',
        desc: '',
      };
    });
  }

  deleteTodo(i) {
    let todos = this.state.todos;
    todos.splice(i, 1);
    this.setState({
      todos,
    });
  }

  onFilterAll() {
    this.setState({ filter: 'all' });
  }

  onFilterActive() {
    this.setState({ filter: 'active' });
  }

  onFilterCompleted() {
    this.setState({ filter: 'completed' });
  }

  onComplete(i) {
    this.setState((prevState) => {
      let todos = prevState.todos.map((todo, index) => {
        todo.completed = index === i ? !todo.completed : todo.completed;
        return todo;
      });
      i = null;
      return {
        todos,
      };
    });
  }

  render() {
    const { name, todos, addTodoActive, updateTodoActive, filter } = this.state;
    let data =
      filter === 'active'
        ? todos.filter((list) => !list.completed)
        : filter === 'completed'
        ? todos.filter((list) => list.completed)
        : todos;
    return (
      <div className='app'>
        {name === '' && (
          <Alert className='alert'>
            <p className='alert__title'>Please Enter your name!</p>
            <form onSubmit={this.onNameSubmit}>
              <input
                type='text'
                className='alert__input'
                onChange={this.onNameChange}
                placeholder='Name'
              />
              <button className='alert__button'>Add</button>
            </form>
          </Alert>
        )}
        {(addTodoActive || updateTodoActive) && (
          <Alert className='alert'>
            <p className='alert__title'>
              {addTodoActive ? `Add` : `Update`} Your New Todo!
            </p>
            <form
              onSubmit={
                addTodoActive ? this.onTodoSubmit : this.onSubmitUpdateTodo
              }>
              <input
                type='text'
                className='alert__input'
                name='title'
                value={this.state.title}
                onChange={this.onTodoChange}
                placeholder='Todo Title'
              />
              <textarea
                className='alert__input'
                rows='4'
                name='desc'
                value={this.state.desc}
                onChange={this.onTodoChange}></textarea>
              <input
                type='submit'
                value={addTodoActive ? `Add` : `Update`}
                className='alert__button'
              />
            </form>
          </Alert>
        )}
        <header>
          <div className='brand'>Machen</div>
          <div className='username'>{name}</div>
        </header>
        <main className='list' id='top'>
          <div className='filters'>
            <button className='filter__all' onClick={this.onFilterAll}>
              All
            </button>
            <button className='filtler__active' onClick={this.onFilterActive}>
              Active
            </button>
            <button
              className='filter__complete'
              onClick={this.onFilterCompleted}>
              Completed
            </button>
          </div>
          {data !== [] ? (
            <ul className='todos__list'>
              {data.map((list, index) => (
                <Todo
                  key={list.id}
                  {...list}
                  index={index}
                  activateUpdateTodo={this.activateUpdateTodo}
                  deleteTodo={this.deleteTodo}
                  onCompleted={() => this.onComplete(index)}
                />
              ))}
            </ul>
          ) : (
            <div className='empty__note'>
              <FontAwesomeIcon icon={faStickyNote} />
            </div>
          )}
          <div className='plus'>
            <FontAwesomeIcon icon={faPlus} onClick={this.activateAddTodo} />
          </div>
          <div className='top'>
            <a href='#top' title='To Top'>
              <FontAwesomeIcon icon={faChevronUp} />
            </a>
          </div>
        </main>
      </div>
    );
  }
}

function Alert({ children }) {
  return <div className='alert'>{children}</div>;
}

export { App, Todo, Alert };
