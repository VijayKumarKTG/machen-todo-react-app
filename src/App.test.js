import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { App, Todo, Alert } from './App';

describe('App', () => {
  it('renders App without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has valid snapshot', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Todo', () => {
  let props = {
    title: 'Stroy Telling',
    desc: 'Tell someone a story',
    completed: true,
    index: 0,
  };

  it('renders Todo without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Todo {...props} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has valid snapshot', () => {
    const component = renderer.create(<Todo {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Alert', () => {
  it('render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Alert>
        <p className='alert__title'>Please Enter your name!</p>
        <form>
          <input type='text' className='alert__input' placeholder='Name' />
          <button className='alert__button'>Add</button>
        </form>
      </Alert>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  test('has valid snapshot', () => {
    const component = renderer.create(
      <Alert>
        <p className='alert__title'>Please Enter your name!</p>
        <form>
          <input type='text' className='alert__input' placeholder='Name' />
          <button className='alert__button'>Add</button>
        </form>
      </Alert>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
