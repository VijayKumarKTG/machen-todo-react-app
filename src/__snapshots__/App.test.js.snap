// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert has valid snapshot 1`] = `
<div
  className="alert"
>
  <p
    className="alert__title"
  >
    Please Enter your name!
  </p>
  <form>
    <input
      className="alert__input"
      placeholder="Name"
      type="text"
    />
    <button
      className="alert__button"
    >
      Add
    </button>
  </form>
</div>
`;

exports[`App has valid snapshot 1`] = `
<div
  className="app"
>
  <div
    className="alert"
  >
    <p
      className="alert__title"
    >
      Please Enter your name!
    </p>
    <form
      onSubmit={[Function]}
    >
      <input
        className="alert__input"
        onChange={[Function]}
        placeholder="Name"
        type="text"
      />
      <button
        className="alert__button"
      >
        Add
      </button>
    </form>
  </div>
  <header>
    <div
      className="brand"
    >
      Machen
    </div>
    <div
      className="username"
    >
      
    </div>
  </header>
  <main
    className="list"
    id="top"
  >
    <div
      className="filters"
    >
      <button
        className="filter__all"
        onClick={[Function]}
      >
        All
      </button>
      <button
        className="filtler__active"
        onClick={[Function]}
      >
        Active
      </button>
      <button
        className="filter__complete"
        onClick={[Function]}
      >
        Completed
      </button>
    </div>
    <ul
      className="todos__list"
    />
    <div
      className="plus"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-plus fa-w-14 "
        data-icon="plus"
        data-prefix="fas"
        focusable="false"
        onClick={[Function]}
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="top"
    >
      <a
        href="#top"
        title="To Top"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-chevron-up fa-w-14 "
          data-icon="chevron-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </div>
  </main>
</div>
`;

exports[`Todo has valid snapshot 1`] = `
<li
  className="list__item complete"
>
  <div
    className="list__layout"
  >
    <span
      className="checkbox"
    >
      <input
        id="complete"
        type="button"
        value=""
      />
    </span>
    <div
      className="list__text"
    >
      <h2
        className="list__item__title"
      >
        Stroy Telling
      </h2>
      <p
        className="list__item__desc"
      >
        Tell someone a story
      </p>
    </div>
    <div
      className="list__icons"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-pencil-alt fa-w-16 icon"
        data-icon="pencil-alt"
        data-prefix="fas"
        focusable="false"
        onClick={[Function]}
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
       
      |
       
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash-alt fa-w-14 icon"
        data-icon="trash-alt"
        data-prefix="fas"
        disabled={true}
        focusable="false"
        onClick={[Function]}
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
  <p
    className="border__bottom"
  />
</li>
`;
